File: Ex2/Ex2.swift
import Foundation

let morseAlphabet: [String] = [
    ".-", "-...", "-.-.", "-..", ".", "..-.", "--.", "....", "..", ".---", "-.-", ".-..", "--", "-.", "---",
    ".--.", "--.-", ".-.", "...", "-", "..-", "...-", ".--", "-..-", "-.--", "--.."
]

// Dictionary to store the mapping between characters and their Morse code representation.
var charToMorse: [Character: String] = [:]

func initCharToMorseMap() {
    for i in 0..<morseAlphabet.count {
        if let scalar = Unicode.Scalar("a".unicodeScalars.first!.value + UInt32(i)) {
            charToMorse[Character(scalar)] = morseAlphabet[i]
        }
    }
}

// Function to generate permutations of the given string and add them to the container.
func generatePermutations(str: String, l: Int, r: Int, permutations: inout Set<String>) {
    var strArray = Array(str)
    if l == r {
        permutations.insert(String(strArray))
    } else {
        for i in l...r {
            strArray.swapAt(l, i)
            generatePermutations(str: String(strArray), l: l + 1, r: r, permutations: &permutations)
            strArray.swapAt(l, i)
        }
    }
}

// Function to convert a string to Morse code using the created map.
func stringToMorse(str: String) -> String {
    var morseString = ""
    for c in str {
        morseString += (charToMorse[c] ?? "")
    }
    return morseString
}

func main() {
    initCharToMorseMap()
    print("Введите слова для перестановки (разделяйте их пробелами): ")
    if let line = readLine() {
        let words = line.components(separatedBy: " ")
        var uniqueMorseWords: Set<String> = []

        for word in words {
            var permutations: Set<String> = []
            generatePermutations(str: word, l: 0, r: word.count - 1, permutations: &permutations)

            for permutation in permutations {
                uniqueMorseWords.insert(stringToMorse(str: permutation))
            }
        }

        print("Уникальные слова в языке Морзе: ")
        for morseWord in uniqueMorseWords {
            print(morseWord)
        }

        print("Количество уникальных слов: \(uniqueMorseWords.count)")
    } else {
        print("Invalid input")
    }
}

main()

File: Ex2/Ex2.cpp
#include <iostream>
#include <algorithm>
#include <vector>
#include <set>
#include <map>
#include <sstream>
using namespace std;

vector<string> morseAlphabet = {
    ".-", "-...", "-.-.", "-..", ".", "..-.", "--.", "....", "..", ".---", "-.-", ".-..", "--", "-.", "---",
    ".--.", "--.-", ".-.", "...", "-", "..-", "...-", ".--", "-..-", "-.--", "--.."
};

// map который будет хранить соответствие между символами и их представлением в коде Морзе.
map<char, string> charToMorse;

void initCharToMorseMap() {
    for (char c = 'a'; c <= 'z'; ++c) {
        charToMorse[c] = morseAlphabet[c - 'a'];
    }
}

// Функция генерирует перестановки переданной строки и добавляет их в контейнер.
void generatePermutations(string str, int l, int r, set<string>& permutations) {
    if (l == r) {
        permutations.insert(str);
    } else {
        for (int i = l; i <= r; i++) {
            swap(str[l], str[i]);
            generatePermutations(str, l + 1, r, permutations);
            swap(str[l], str[i]);
        }
    }
}

// Преобразую строку в код Морзе, используя созданный map.
string stringToMorse(const string& str) {
    string morseString = "";
    for (char c : str) {
        morseString += charToMorse[c];
    }
    return morseString;
}

int main() {
    int count = 0;
    initCharToMorseMap();
    setlocale(LC_ALL, "Russian");
    string input;
    cout << "Введите слова для перестановки (разделяйте их пробелами): ";
    getline(cin, input);

    // Разделяем ввод на отдельные слова.
    stringstream ss(input);
    string word;
    set<string> uniqueMorseWords;

    while (ss >> word) {
        set<string> permutations;
        generatePermutations(word, 0, word.size() - 1, permutations);

        for (const auto& permutation : permutations) {
            uniqueMorseWords.insert(stringToMorse(permutation));
        }
    }

    cout << "Уникальные слова в языке Морзе: ";
    for (const auto& morseWord : uniqueMorseWords) {
        count++;
        cout << morseWord << endl;
    }
     cout << "Количество уникальных слов: " << count << endl;
    return 0;
}